### PR TITLE
Rule S1075: URIs should not be hardcoded

### LINKED ISSUE #507: Rule S1075: URIs should not be hardcoded
[RSPEC-1075](https://jira.sonarsource.com/browse/RSPEC-1075)

Hard coding a URI makes it difficult to test a program: path literals are not always portable across operating systems, a given absolute path may not exist on a specific test environment, a specified Internet URL may not be available when executing the tests, production environment filesystems usually differ from the development environment, ...etc. For all those reasons, a URI should never be hard coded. Instead, it should be replaced by customizable parameter.

Contributes to [MMF-963](https://jira.sonarsource.com/browse/MMF-963).

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #507

### CODE REVIEW

**Comment by Evangelink** [comment]
I haven't reviewed all raised issues yet but will definitely do it tomorrow.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                    .FirstOrDefault(ancestor => ancestor.IsKind(SyntaxKind.InvocationExpression) ||
+                                                ancestor.IsKind(SyntaxKind.ObjectCreationExpression));
+                var methodSymbol = constructorOrMethod != null
+                    ? model.GetSymbolInfo(constructorOrMethod).Symbol as IMethodSymbol
+                    : null;
+
+                return (methodSymbol != null && argumentIndex.Value < methodSymbol.Parameters.Length)
+                    ? methodSymbol.Parameters[argumentIndex.Value].Name.SplitCamelCaseToWords()
+                        .Any(name => checkedVariableNames.Contains(name))
+                    : false;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            }
+
+            var variableDeclarator = expression.FirstAncestorOrSelf<VariableDeclaratorSyntax>();
+
+            return variableDeclarator != null
+                ? variableDeclarator.Identifier.ValueText
+                    .SplitCamelCaseToWords()
+                    .Any(name => checkedVariableNames.Contains(name))
+                : false;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 2 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 2 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class UriShouldNotBeHardcoded : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S1075";
+        private const string MessageFormat = "{0}";
+        private const string AbsoluteUriMessage = "Refactor your code to get this URI from a customizable parameter.";
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: I don't like this message. "customizable parameter" sounds weird. Could parameter be not customizable?
Perhaps: "Refactor your code not to use hardcoded path or URI."?

Also, I prefer "hardcoded" over "hard-coded" everywhere.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+                    : false;
+            }
+
+            var variableDeclarator = expression.FirstAncestorOrSelf<VariableDeclaratorSyntax>();
+
+            return variableDeclarator != null
+                ? variableDeclarator.Identifier.ValueText
+                    .SplitCamelCaseToWords()
+                    .Any(name => checkedVariableNames.Contains(name))
+                : false;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+                    .FirstOrDefault(ancestor => ancestor.IsKind(SyntaxKind.InvocationExpression) ||
+                                                ancestor.IsKind(SyntaxKind.ObjectCreationExpression));
+                var methodSymbol = constructorOrMethod != null
+                    ? model.GetSymbolInfo(constructorOrMethod).Symbol as IMethodSymbol
+                    : null;
+
+                return (methodSymbol != null && argumentIndex.Value < methodSymbol.Parameters.Length)
+                    ? methodSymbol.Parameters[argumentIndex.Value].Name.SplitCamelCaseToWords()
+                        .Any(name => checkedVariableNames.Contains(name))
+                    : false;
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 4 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs (hunk 5 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/UriShouldNotBeHardcoded.cs]: MINOR Remove the unnecessary Boolean literal(s). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1125)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**